src/main/java/com/dropbox/core/android/AuthActivity.java

```diff
@@ -581,4 +675,56 @@ private String createStateNonce() {
         }
         return sb.toString();
     }
+
+    private String createPKCEStateNonce() {
+        return String.format("oauth2code:%s:%s:%s",
+                             mPKCEManager.getCodeChallenge(),
+                             DbxPKCEManager.CODE_CHALLENGE_METHODS,
+                             mTokenAccessType.toString());
+    }
+
+    private String createExtraQueryParams() {
+        return String.format(
+            "%s=%s&%s=%s&%s=%s&%s=%s",
+            "code_challenge", mPKCEManager.getCodeChallenge(),
+            "code_challenge_method", DbxPKCEManager.CODE_CHALLENGE_METHODS,
+            "token_access_type", mTokenAccessType.toString(),
+            "response_type", "code"
+        );
+    }
+
+    private enum TokenType {
+        OAUTH2("oauth2:"),
+        OAUTH2CODE("oauth2code:");
+
+        private String string;
+
+        TokenType(String string) {
+            this.string = string;
+        }
+
+        @Override
+        public String toString() {
+            return string;
+        }
+    }
+
+    private class TokenRequest extends AsyncTask<Void, Void, DbxAuthFinish> {
+        private final String code;
+
+        private TokenRequest(String code) {
+            this.code = code;
+        }
+
+
+        @Override
+        protected DbxAuthFinish doInBackground(Void... p) {
+            try {
+                return mPKCEManager.makeTokenRequest(mRequestConfig, code, mAppKey, null,mHost);
```

Comment:
We have to perform an async task here to finish OAuth code flow. I understand its not a good practice. But this solution guarantee backward compatibility.

Alternatively, we can return this task as a callback to 3rd party apps and let them finish this call in back ground. But that's a bigger cost for migration.

Reply:
nit: missing space.